cocos/platform/CCFileUtils.cpp

```diff
@@ -817,10 +824,14 @@ std::string FileUtils::fullPathForFilename(const std::string &filename) const
     {
         return filename;
     }
+    //dead lock check
+    auto fullPathCache = _fullPathCache.load();
```

Comment:
1. Yes. Currently it is very easy lock mutexes in wrong order and have dead lock ...
2. I think that instead of 6 mutexes (Atomic mutexes) we should create one mutex inside FileUtils and lock it when needed (on beginning of each function where there is access to member data? )

Reply:
I think @PatriceJiang 's design is good, it makes things easy.

Reply:
Please check following link: https://wandbox.org/nojs/gcc-8.2.0/permlink/ASbHxrzF3j6EQJdD
And compare implementation of struct Counter and AtomicCounter. I think that class/struct with lock in each public method is much better in this case.
1. struct Counter is much simpler than AtomicCounter
2. struct Counter is 20x faster than AtomicCounter
3. I think that usage of std::recursive_mutex is bad design here
Please check qoute from: "Cplusplus Concurrency In Action"

>  Most of the time, if you think you want a recursive mutex, you probably need to
> change your design instead. A common use of recursive mutexes is where a class is
> designed to be accessible from multiple threads concurrently, so it has a mutex protecting
> the member data. Each public member function locks the mutex, does the
> work, and then unlocks the mutex. However, sometimes it’s desirable for one public
> member function to call another as part of its operation. In this case, the second
> member function will also try to lock the mutex, thus leading to undefined behavior.
> The quick-and-dirty solution is to change the mutex to a recursive mutex. This
> will allow the mutex lock in the second member function to succeed and the function
> to proceed.
>  However, such usage is not recommended, because it can lead to sloppy thinking
> and bad design. In particular, the class invariants are typically broken while the lock
> is held, which means that the second member function needs to work even when
> called with the invariants broken. It’s usually better to extract a new private member
> function that’s called from both member functions, which does not lock the mutex
> (it expects it to already be locked). You can then think carefully about the circumstances
> under which that new function can be called and the state of the data under
> those circumstances

Reply:
`recursive_mutex` is used to support 
```cpp
{
  Atomic a;
  auto x = a.load();
  {
 //...
      auto y = a.load(); // or invoked in other function 
  //...
  }
}
```

Reply:
1. "we don't have to modify all datas at the same time"
In almost each function we protect _fullPathCache
from the first a.load() all data are protected
2. More than 2 mutex locked in different order == dead lock
In this code 6 mutex were introduced. For example below code has dead lock (a1, a2 order  different)
```
void f1()
{
    a1.load();
    a2.load();
}  
void f2()
{
    a2.load();
    a1.load();
}
```
Are you sure that **always** all atomics variable are loaded (mutex is locked) in the same order?
a1.load()
a2.load()
a3.load()
....
Please remember  that currently in each function with load ( mutex lock) may call another function with load ( mutex lock) and so on.... Are you 100% sure that all loads are always in the same order for **all possible execution path**? (I checked and I am not sure.... )
Even if we check that in current code for all possible path all load are in the same order I think that sooner or later someone will write another member function when order will be broken (e.g. a2 load() id before a1.load() and we have dead lock.......
p.s. It is really hard to reproduce dead lock :(

Reply:
current locks are arranged simply in alphabet order of field names.

Reply:
Please check below two functions. "locks are arranged simply in alphabet order of field names" inside each function. But if you have functions in different files (android/win32) and basically from any function you can call any other function you can very easy create dead lock.

void f1()
{
    a2.load();
    // many lines of code
    f2();
} 

// different file
void f2()
{
    a1.load();
}

Reply:
@crazyhappygame please narrow down the discussion scope. We just discuss the file CCFileUtils.cpp, it doesn't across files.

Reply:
In this code review there is CCFileUtils-win32.cpp, CCFileUtils-android.cpp. :)

Reply:
But let me rephrase. Are we 100% sure that in current code we do not reorder locks like below?
void f1()
{
a2.load();
f2();
}

void f2()
{
a1.load();
}

Reply:
Oh, yep, if across functions, it is very difficult to make sure the order.